File: book/스프링부트를 활용한 마이크로서비스 개발/test/src/test/java/com/example/test/service/MultiplicationServiceImplTest.java
package com.example.test.service;

import com.example.test.domain.Multiplication;
import com.example.test.domain.MultiplicationResultAttempt;
import com.example.test.domain.User;
import com.example.test.repository.MultiplicationResultAttemptRepository;
import com.example.test.repository.UserRepository;
import org.assertj.core.api.AssertionsForClassTypes;
import org.assertj.core.util.Lists;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import javax.swing.text.html.Option;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.verify;

public class MultiplicationServiceImplTest {

    private MultiplicationServiceImpl multiplicationServiceImpl;

    @Mock
    private RandomGeneratorService randomGeneratorService;

    @Mock
    private MultiplicationResultAttemptRepository attemptRepository;

    @Mock
    private UserRepository userRepository;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);
        multiplicationServiceImpl = new MultiplicationServiceImpl(randomGeneratorService
        , userRepository, attemptRepository);
    }

    @Test
    public void createRandomMultiplication(){
        // given
        given(randomGeneratorService.generateFactor()).willReturn(50, 30);

        Multiplication multiplication = multiplicationServiceImpl.createRandomMultiplication();

        AssertionsForClassTypes.assertThat(multiplication.getFactorA()).isEqualTo(50);
        AssertionsForClassTypes.assertThat(multiplication.getFactorB()).isEqualTo(30);
//        AssertionsForClassTypes.assertThat(multiplication.getResult()).isEqualTo(1500);
    }

    @Test
    public void checkCorrectAttemptTest() {
        Multiplication multiplication = new Multiplication(50, 60);
        User user = new User("John");
        MultiplicationResultAttempt attempt = new MultiplicationResultAttempt(user,
                multiplication, 3000, false);
        MultiplicationResultAttempt verifiedAttempt = new MultiplicationResultAttempt(user,
                multiplication, 3000, true);
        given(userRepository.findByAlias("John")).willReturn(Optional.empty());

        boolean attemptResult = multiplicationServiceImpl.checkAttempt(attempt);

        assertThat(attemptResult).isTrue();
        verify(attemptRepository).save(verifiedAttempt);
    }

    @Test
    public void checkWrongAttemptTest() {
        Multiplication multiplication = new Multiplication(50, 60);
        User user = new User("John");
        MultiplicationResultAttempt attempt = new MultiplicationResultAttempt(user,
                multiplication, 3010, false);
        given(userRepository.findByAlias("John")).willReturn(Optional.empty());

        boolean attemptResult = multiplicationServiceImpl.checkAttempt(attempt);

        assertThat(attemptResult).isFalse();
        verify(attemptRepository).save(attempt);
    }

    @Test
    public void retrieveStatsTest() {
        Multiplication multiplication = new Multiplication(50, 60);
        User user = new User("John");
        MultiplicationResultAttempt attempt1 = new MultiplicationResultAttempt(user,
                multiplication, 3010, false);
        MultiplicationResultAttempt attempt2 = new MultiplicationResultAttempt(user,
                multiplication, 3051, false);
        List<MultiplicationResultAttempt> latestAttempts = Lists.newArrayList(attempt1, attempt2);
        given(userRepository.findByAlias("John")).willReturn(Optional.empty());
        given(attemptRepository.findTop5ByUserAliasOrderByIdDesc("John")).willReturn(latestAttempts);

        List<MultiplicationResultAttempt> latestAttemptsResult =
                multiplicationServiceImpl.getStatsForUser("John");

        assertThat(latestAttemptsResult).isEqualTo(latestAttempts);

    }

}

File: book/스프링부트를 활용한 마이크로서비스 개발/kotlin/src/test/kotlin/com/example/kotlin/service/MultiplicationServiceImplTest.kt
package com.example.kotlin.service

import com.example.kotlin.domain.Multiplication
import com.example.kotlin.domain.MultiplicationResultAttempt
import com.example.kotlin.domain.User
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.mockito.BDDMockito
import org.mockito.Mock
import org.mockito.MockitoAnnotations

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class MultiplicationServiceImplTest {

    private lateinit var multiplicationServiceImpl : MultiplicationServiceImpl

    @Mock
    lateinit var randomGeneratorService : RandomGeneratorService

    // @Before -> JUnit4
    // @BeforeAll -> JUnit5 -> 근데 이거 쓰려면 자바에서도 static 설정을 해줘야 한다는데
    // 코틀린에서는 위에 @TestInstance(Lifecycle.PER_CLASS) 안하면 에러 뜬다.
    // 이렇게 해두면 static 메서드가 아닌 곳에서도 사용 가능하다고.(테스트 클래스 당 인스턴스를 생성해서)
    @BeforeAll
    fun setup() {
        MockitoAnnotations.initMocks(this)
        multiplicationServiceImpl = MultiplicationServiceImpl(randomGeneratorService)
    }

    @Test
    fun createRandomMultiplicationTest() {
        // given (randomGeneratorService 가 처음에는 50, 나중에는 30을 반환하게 설정)
        BDDMockito.given(randomGeneratorService.generateRandomFactor()).willReturn(50, 30)

        // when
        val multiplication = multiplicationServiceImpl.createRandomMultiplication()

        // assert
        assertThat(multiplication.factorA).isEqualTo(50)
        assertThat(multiplication.factorB).isEqualTo(30)
        assertThat(multiplication.result).isEqualTo(1500)
    }

    @Test
    fun checkCorrectAttemptTest() {
        // given
        val multiplication = Multiplication(50, 60)
        val user = User("seoin")
        val attempt = MultiplicationResultAttempt(user, multiplication, 3000)

        // when
        val attemptResult = multiplicationServiceImpl.checkAttempt(attempt)

        // assert
        assertThat(attemptResult).isTrue
    }

    @Test
    fun checkWrongAttemptTest() {
        // given
        val multiplication = Multiplication(50, 60)
        val user = User("seoin")
        val attempt = MultiplicationResultAttempt(user, multiplication, 3010)

        // when
        val attemptResult = multiplicationServiceImpl.checkAttempt(attempt)

        // assert
        assertThat(attemptResult).isFalse
    }
}